File: backend/app/schemas/scorecard.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


# ── Indicator Values ──

class IndicatorValueCreate(BaseModel):
    year: int
    period_type: str = "monthly"  # monthly | quarterly | annual
    period: int  # 1-12 monthly, 1-4 quarterly, 0 annual
    value: float
    collection_method: str = "manual"
    notes: str | None = None
    actions_done: str | None = None
    actions_next: str | None = None


class IndicatorValueUpdate(BaseModel):
    value: float | None = None
    notes: str | None = None
    actions_done: str | None = None
    actions_next: str | None = None


class IndicatorValueRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    indicator_id: uuid.UUID
    year: int
    period_type: str
    period: int
    value: float
    collected_at: datetime
    collection_method: str
    notes: str | None = None
    actions_done: str | None = None
    actions_next: str | None = None


# ── Scorecard Indicators ──

class ScorecardIndicatorCreate(BaseModel):
    name: str
    description: str | None = None
    category: str  # operational | management
    indicator_type: str = "continuous"  # continuous | binary | count
    target_value: float
    unit: str = "%"
    weight_pct: float = 16.67
    direction: str = "higher_better"
    frequency: str = "monthly"
    start_date: date | None = None
    end_date: date | None = None
    source: str | None = None
    source_config: dict | None = None
    kpi_definition_id: uuid.UUID | None = None
    project_id: uuid.UUID | None = None
    connector_id: uuid.UUID | None = None
    shared_group_id: uuid.UUID | None = None
    order_idx: int = 0


class ScorecardIndicatorUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    category: str | None = None
    indicator_type: str | None = None
    target_value: float | None = None
    unit: str | None = None
    weight_pct: float | None = None
    direction: str | None = None
    frequency: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    source: str | None = None
    source_config: dict | None = None
    kpi_definition_id: uuid.UUID | None = None
    project_id: uuid.UUID | None = None
    connector_id: uuid.UUID | None = None
    shared_group_id: uuid.UUID | None = None
    order_idx: int | None = None


class ScorecardIndicatorRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    scorecard_id: uuid.UUID
    name: str
    description: str | None = None
    category: str
    indicator_type: str
    target_value: float
    unit: str
    weight_pct: float
    direction: str
    frequency: str
    start_date: date | None = None
    end_date: date | None = None
    source: str | None = None
    source_config: dict | None = None
    kpi_definition_id: uuid.UUID | None = None
    project_id: uuid.UUID | None = None
    connector_id: uuid.UUID | None = None
    shared_group_id: uuid.UUID | None = None
    order_idx: int
    active: bool
    values: list[IndicatorValueRead] = []
    # Computed fields (filled by service)
    current_value: float | None = None
    progress: float | None = None
    status: str | None = None


# ── Scorecards ──

class ScorecardCreate(BaseModel):
    company_id: uuid.UUID
    year: int
    org_level: str  # department | area | team
    org_entity_id: uuid.UUID
    name: str
    operational_weight: float = 50.0
    management_weight: float = 50.0


class ScorecardUpdate(BaseModel):
    name: str | None = None
    operational_weight: float | None = None
    management_weight: float | None = None


class ScorecardRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_id: uuid.UUID
    year: int
    org_level: str
    org_entity_id: uuid.UUID
    name: str
    operational_weight: float
    management_weight: float
    active: bool
    indicators: list[ScorecardIndicatorRead] = []
    # Computed fields (filled by service)
    operational_score: float | None = None
    management_score: float | None = None
    overall_score: float | None = None
    org_entity_name: str | None = None
    responsible_name: str | None = None


class ScorecardTreeNode(BaseModel):
    """A node in the hierarchical scorecard tree."""
    scorecard: ScorecardRead | None = None
    org_level: str
    org_entity_id: uuid.UUID
    org_entity_name: str
    responsible_name: str | None = None
    overall_score: float | None = None
    children: list["ScorecardTreeNode"] = []

File: frontend/src/types/scorecard.ts
export interface IndicatorValue {
  id: string;
  indicator_id: string;
  year: number;
  period_type: "monthly" | "quarterly" | "annual";
  period: number;
  value: number;
  collected_at: string;
  collection_method: string;
  notes: string | null;
  actions_done: string | null;
  actions_next: string | null;
}

export interface IndicatorValueCreate {
  year: number;
  period_type: "monthly" | "quarterly" | "annual";
  period: number;
  value: number;
  collection_method?: string;
  notes?: string;
  actions_done?: string;
  actions_next?: string;
}

export interface ScorecardIndicator {
  id: string;
  scorecard_id: string;
  name: string;
  description: string | null;
  category: "operational" | "management";
  indicator_type: "continuous" | "binary" | "count";
  target_value: number;
  unit: string;
  weight_pct: number;
  direction: "higher_better" | "lower_better";
  frequency: "monthly" | "quarterly" | "annual";
  source: string | null;
  source_config: Record<string, unknown> | null;
  kpi_definition_id: string | null;
  project_id: string | null;
  connector_id: string | null;
  shared_group_id: string | null;
  order_idx: number;
  active: boolean;
  values: IndicatorValue[];
  // Computed
  current_value: number | null;
  progress: number | null;
  status: "ok" | "warning" | "ko" | "no_data" | null;
}

export interface Scorecard {
  id: string;
  company_id: string;
  year: number;
  org_level: "department" | "area" | "team";
  org_entity_id: string;
  name: string;
  operational_weight: number;
  management_weight: number;
  active: boolean;
  indicators: ScorecardIndicator[];
  // Computed
  operational_score: number | null;
  management_score: number | null;
  overall_score: number | null;
  org_entity_name: string | null;
  responsible_name: string | null;
}

export interface ScorecardTreeNode {
  org_level: "department" | "area" | "team";
  org_entity_id: string;
  org_entity_name: string;
  responsible_name: string | null;
  overall_score: number | null;
  operational_score: number | null;
  management_score: number | null;
  scorecard: Scorecard | null;
  children: ScorecardTreeNode[];
}
